### DISCUSSION TITLE
heltec_wifi_lora_32_V3 - radiolib - CayenneLPP

### DISCUSSION THREAD

**Discussion by tanguy-e** [jgromes/RadioLib#979] [Q&A] [answered]
Hello, i'm trying to use Radiolib and LoRaWAN-End-Device example with heltec_wifi_lora_32_V3, BMP280 and Cayenne.
The base example works fine the "Hello World ! #N" is received by the TTN application. 
After this i tried to modify the example : 
void loop() {
  float temperature = bmp.readTemperature();
  float pressure = bmp.readPressure();

  lpp.reset();
  lpp.addTemperature(1, temperature); // Add temperature data to channel 1
  lpp.addBarometricPressure(2, pressure); // Add pressure data to channel 2
  byte packet[lpp.getSize()];
  lpp.copy(packet); // Copy the CayenneLPP packet to the buffer
  // send uplink to port 10
  Serial.print(F("[LoRaWAN] Sending uplink packet ... "));
  String strDown;
  node.sendReceive(lpp.getBuffer(), 10, strDown);

But the "node.sendReceive( ...);" seems to lead to compilation error.
Could you help me to understand the error and correct the mistake ?
Thanks

**Comment by jgromes** [comment]
> But the "node.sendReceive( ...);" seems to lead to compilation error.
> Could you help me to understand the error and correct the mistake ?

Not without seeing the actual error.

I'm guessing that `lpp.getBuffer()` returns some `uint8_t *`, which would mean you are trying to call `sendReceive(uint8_t*, int, String&)`, which doesn't exist - see the API reference linked from the readme to find out which types `sendReceive` supports. Just an educated guess though.

**Reply by tanguy-e** [reply]
It seems you're right : 
```
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino: In function 'void loop()':
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:141:45: error: invalid user-defined conversion from 'uint8_t*' {aka 'unsigned char*'} to 'String&' [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), 10, strDown);
                                ~~~~~~~~~~~~~^~
In file included from C:\Users\tanguy-e\AppData\Local\Arduino15\packages\Heltec-esp32\hardware\esp32\1.0.0\cores\esp32/Arduino.h:165,
                 from C:\Users\tanguy-e\AppData\Local\Temp\arduino\sketches\27A11A429D7ABFC4AB8829B005DBADD7\sketch\LoRaWAN_End_Device.ino.cpp:1:
C:\Users\tanguy-e\AppData\Local\Arduino15\packages\Heltec-esp32\hardware\esp32\1.0.0\cores\esp32/WString.h:57:9: note: candidate is: 'String::String(const char*)' <near match>
         String(const char *cstr = "");
         ^~~~~~
C:\Users\tanguy-e\AppData\Local\Arduino15\packages\Heltec-esp32\hardware\esp32\1.0.0\cores\esp32/WString.h:57:9: note:   conversion of argument 1 would be ill-formed:
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:141:45: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'const char*' [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), 10, strDown);
                                ~~~~~~~~~~~~~^~
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:141:45: error: invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'const char*' [-fpermissive]
In file included from C:\Users\tanguy-e\AppData\Local\Arduino15\packages\Heltec-esp32\hardware\esp32\1.0.0\cores\esp32/Arduino.h:165,
                 from C:\Users\tanguy-e\AppData\Local\Temp\arduino\sketches\27A11A429D7ABFC4AB8829B005DBADD7\sketch\LoRaWAN_End_Device.ino.cpp:1:
C:\Users\tanguy-e\AppData\Local\Arduino15\packages\Heltec-esp32\hardware\esp32\1.0.0\cores\esp32/WString.h:57:28: note:   initializing argument 1 of 'String::String(const char*)'
         String(const char *cstr = "");
                ~~~~~~~~~~~~^~~~~~~~~
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:141:60: error: conversion to non-const reference type 'class String&' from rvalue of type 'String' [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), 10, strDown);
                                                            ^
In file included from c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/RadioLib.h:117,
                 from C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:25:
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:546:13: note:   initializing argument 1 of 'int16_t LoRaWANNode::sendReceive(String&, uint8_t, String&, bool, LoRaWANEvent_t*, LoRaWANEvent_t*)'
     int16_t sendReceive(String& strUp, uint8_t port, String& strDown, bool isConfirmed = false, LoRaWANEvent_t* eventUp = NULL, LoRaWANEvent_t* eventDown = NULL);
             ^~~~~~~~~~~

exit status 1

Compilation error: invalid user-defined conversion from 'uint8_t*' {aka 'unsigned char*'} to 'String&' [-fpermissive]
```

but it seems that is it possible to use this with uint8_t * : 
sendReceive | ( | uint8_t * | dataUp,
  |   | size_t | lenUp,
  |   | uint8_t | port,
  |   | uint8_t * | dataDown,
  |   | size_t * | lenDown,
  |   | bool | isConfirmed = false,
  |   | LoRaWANEvent_t * | eventUp = NULL,
  |   | LoRaWANEvent_t * | eventDown = NULL
  | )
i'm not sure how to do this ....

**Comment by tanguy-e** [comment]
I tried : 

```c++
  void loop() {
  float temperature = bmp.readTemperature();
  float pressure = bmp.readPressure();

  lpp.reset();
  lpp.addTemperature(1, temperature); // Add temperature data to channel 1
  lpp.addBarometricPressure(2, pressure); // Add pressure data to channel 2

  byte packet[lpp.getSize()];
  lpp.copy(packet); // Copy the CayenneLPP packet to the buffer

  Serial.print(F("[LoRaWAN] Sending uplink packet ... "));
  String strUp = "Hello World! #" + String(count++);
  int strDown;
  int state = node.sendReceive(lpp.getBuffer(), lpp.getSize(), 10, strDown, 10);
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("received a downlink!"));
```
without success!
There is still a type mismatch that i don't understand : 

```
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino: In function 'void loop()':
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:142:79: error: no matching function for call to 'sendReceive(uint8_t*, uint8_t, int, int&, int)'
   int state = node.sendReceive(lpp.getBuffer(), lpp.getSize(), 10, strDown, 10);
                                                                               ^
In file included from c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/RadioLib.h:117,
                 from C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:25:
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:546:13: note: candidate: 'int16_t LoRaWANNode::sendReceive(String&, uint8_t, String&, bool, LoRaWANEvent_t*, LoRaWANEvent_t*)' <near match>
     int16_t sendReceive(String& strUp, uint8_t port, String& strDown, bool isConfirmed = false, LoRaWANEvent_t* eventUp = NULL, LoRaWANEvent_t* eventDown = NULL);
             ^~~~~~~~~~~
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:546:13: note:   conversion of argument 5 would be ill-formed:
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:142:77: error: invalid conversion from 'int' to 'LoRaWANEvent_t*' [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), lpp.getSize(), 10, strDown, 10);
                                                                             ^~
In file included from c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/RadioLib.h:117,
                 from C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:25:
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:562:13: note: candidate: 'int16_t LoRaWANNode::sendReceive(const char*, uint8_t, uint8_t*, size_t*, bool, LoRaWANEvent_t*, LoRaWANEvent_t*)' <near match>
     int16_t sendReceive(const char* strUp, uint8_t port, uint8_t* dataDown, size_t* lenDown, bool isConfirmed = false, LoRaWANEvent_t* eventUp = NULL, LoRaWANEvent_t* eventDown = NULL);
             ^~~~~~~~~~~
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:562:13: note:   conversion of argument 4 would be ill-formed:
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:142:68: error: invalid conversion from 'int' to 'size_t*' {aka 'unsigned int*'} [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), lpp.getSize(), 10, strDown, 10);
                                                                    ^~~~~~~
In file included from c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/RadioLib.h:117,
                 from C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:25:
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:578:13: note: candidate: 'int16_t LoRaWANNode::sendReceive(uint8_t*, size_t, uint8_t, uint8_t*, size_t*, bool, LoRaWANEvent_t*, LoRaWANEvent_t*)' <near match>
     int16_t sendReceive(uint8_t* dataUp, size_t lenUp, uint8_t port, uint8_t* dataDown, size_t* lenDown, bool isConfirmed = false, LoRaWANEvent_t* eventUp = NULL, LoRaWANEvent_t* eventDown = NULL);
             ^~~~~~~~~~~
c:\Users\tanguy-e\Documents\Arduino\libraries\RadioLib\src/protocols/LoRaWAN/LoRaWAN.h:578:13: note:   conversion of argument 5 would be ill-formed:
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:142:77: error: invalid conversion from 'int' to 'size_t*' {aka 'unsigned int*'} [-fpermissive]
   int state = node.sendReceive(lpp.getBuffer(), lpp.getSize(), 10, strDown, 10);
                                                                             ^~
C:\Users\tanguy-e\Documents\Arduino\LoRaWAN_End_Device\LoRaWAN_End_Device.ino:149:16: error: request for member 'length' in 'strDown', which is of non-class type 'int'
     if(strDown.length() > 0) {
                ^~~~~~

exit status 1

Compilation error: no matching function for call to 'sendReceive(uint8_t*, uint8_t, int, int&, int)'
```

**Reply by jgromes** [reply]
The compiler tells you exactly what is going wrong: there is no function that matches the signature `sendReceive(uint8_t*, uint8_t, int, int&, int)`, you just changed `String strDown;` to `int strDown;`, I don't understand what you were trying to do there.

You have exactly three options for `sendReceive` (optional arguments left out for brevity):

* `sendReceive(String&, uint8_t, String&);`
* `sendReceive(const char* strUp, uint8_t port, uint8_t* dataDown, size_t* lenDown);`
* `sendReceive(uint8_t* dataUp, size_t lenUp, uint8_t port, uint8_t* dataDown, size_t* lenDown);`

Yours is neither of those. If you want to do something else, you can always call `uplink` and `downlink` manually, that's how `sendReceive` is implemeted. It's very simple:

```c++
int state = node.uplink(lpp.getBuffer(), lpp.getSize(), 10);
if(state != RADIOLIB_ERR_NONE) {
  // error handling here
}
String strDown;
state = node.downlink(strDown);
if(state != RADIOLIB_ERR_NONE) {
  // error handling here
}
```

**Reply by tanguy-e** [reply]
Thanks you're right i'm mixing up the wordings. I will give a try to use uplink and downlink but i don't have the card with me at this time. I'm quite a newbie and i try to replace what i did with the Heltec LoRaWAN with radiolib. I need to test more and try different things and I'm definitely going to have new questions...

**Comment by tanguy-e** [comment]
Now i have something that works fine : 

#include <RadioLib.h>                 // Inclusion de la bibliothèque RadioLib pour la communication radio
#include <Wire.h>                     // Inclusion de la bibliothèque Wire pour la communication I2C
#include <Adafruit_BMP280.h>          // Inclusion de la bibliothèque Adafruit_BMP280 pour le capteur de pression et de température
#include <CayenneLPP.h>               // Inclusion de la bibliothèque CayenneLPP pour le formatage des données pour LoRaWAN

SX1262 radio = new Module(8, 14, 12, 13);   // Création d'une instance de module LoRa avec les broches spécifiées

LoRaWANNode node(&radio, &EU868);            // Création d'une instance de nœud LoRaWAN avec le module radio et la bande de fréquence spécifiée

Adafruit_BMP280 bmp;                         // Création d'une instance de capteur BMP280
CayenneLPP lpp(51);                          // Création d'une instance de CayenneLPP avec une taille maximale de 51 octets

void setup() {
  Serial.begin(112500);                      // Initialisation de la communication série à 112500 bauds
  
  unsigned status;                            // Déclaration d'une variable pour le statut
  status = bmp.begin(BMP280_ADDRESS_ALT, BMP280_CHIPID);   // Initialisation du capteur BMP280 avec l'adresse et l'ID spécifiés
  if (!status) {
    Serial.println(F("Could not find a valid BMP280 sensor, check wiring or "
                     "try a different address!"));   // Message d'erreur si le capteur BMP280 n'est pas trouvé
  }
  bmp.setSampling(Adafruit_BMP280::MODE_NORMAL,     
                  Adafruit_BMP280::SAMPLING_X2,     
                  Adafruit_BMP280::SAMPLING_X16,    
                  Adafruit_BMP280::FILTER_X16,      
                  Adafruit_BMP280::STANDBY_MS_500);   // Configuration de l'échantillonnage du capteur BMP280

  int state = radio.begin();                  // Initialisation du module LoRa
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));            // Message de réussite de l'initialisation
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  uint64_t joinEUI = 0xXX;     // Identifiant d'application pour la connexion OTAA
  uint64_t devEUI = 0xXX;       // Identifiant de périphérique pour la connexion OTAA

  uint8_t nwkKey[] = {XX  };   // Clé de réseau pour la connexion OTAA
  uint8_t appKey[] = { XX };   // Clé d'application pour la connexion OTAA

  Serial.print(F("[LoRaWAN] Tentative de connexion OTAA ... "));
  state = node.beginOTAA(joinEUI, devEUI, nwkKey, appKey);   // Connexion au réseau LoRaWAN via la méthode OTAA

  if(state >= RADIOLIB_ERR_NONE) {
    Serial.println(F("Connexion!"));          // Message de réussite de la connexion
    delay(2000);                              // Petit délai entre la connexion et l'envoi de données
  } else {
    Serial.print(F("Echoué, code "));
    Serial.println(state);
    while(true);
  }
}

void loop() {
  char str[30];                               // Déclaration d'un tableau de caractères pour stocker des chaînes de caractères
  float temperature = bmp.readTemperature();   // Lecture de la température à partir du capteur BMP280
  float pressure = (bmp.readPressure()/100.0); // Lecture de la pression atmosphérique à partir du capteur BMP280 et conversion en hPa

  sprintf(str,"Température : %.1f °C",temperature);   // Formatage de la température dans une chaîne de caractères
  Serial.println();
  Serial.println(str);                                 // Affichage de la température sur le moniteur série
  sprintf(str,"Pression : %.1f °hPa",pressure);        // Formatage de la pression dans une chaîne de caractères
  Serial.println(str);                                 // Affichage de la pression sur le moniteur série
  
  lpp.reset();                                          // Réinitialisation de l'objet CayenneLPP

  lpp.addTemperature(1, temperature);                  // Ajout de la température au paquet CayenneLPP sur le canal 1
  lpp.addBarometricPressure(2, pressure);              // Ajout de la pression au paquet CayenneLPP sur le canal 2

  byte packet[lpp.getSize()];                          // Déclaration d'un tableau d'octets pour stocker le paquet CayenneLPP
  lpp.copy(packet);                                    // Copie du paquet CayenneLPP dans le tableau d'octets

  Serial.print("Octets envoyés : ");
  Serial.println(lpp.getSize());                       // Affichage du nombre d'octets envoyés

  Serial.print("Trame envoyée : ");
  uint8_t *payload = lpp.getBuffer();

  for (unsigned char i = 0; i < lpp.getSize(); i++)   // Boucle pour afficher chaque octet de la trame envoyée en format hexadécimal
  {
    Serial.print("0x");
    Serial.print(payload[i], HEX);
    Serial.print(" ");
  }
  Serial.println();

  Serial.println(F("[LoRaWAN] En train d'envoyer la trame ... "));
  int state = node.uplink(lpp.getBuffer(), lpp.getSize(), 10);   // Envoi du paquet CayenneLPP en tant que transmission ascendante
    String strDown;  // Déclaration d'une variable pour stocker les données de la réception descendante

  if(state != RADIOLIB_ERR_NONE) {
    Serial.println(F("Erreur d'envoi!"));  // Message d'erreur en cas d'échec de l'envoi
  }

  state = node.downlink(strDown);  // Réception des données descendantes

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Réception reçue!"));  // Message de réussite de la réception

    Serial.println(F("[LoRaWAN] Data:\t\t"));
    if(strDown.length() > 0) {
      Serial.println(strDown);  // Affichage des données de la réception si elles existent
    } else {
      Serial.println(F("<MAC commands only>"));  // Affichage si seules des commandes MAC sont reçues
      Serial.println(strDown);
    }

    Serial.print(F("[LoRaWAN] RSSI:\t\t"));
    Serial.print(radio.getRSSI());  // Affichage de la force du signal RSSI

    Serial.println(F(" dBm"));

    Serial.print(F("[LoRaWAN] SNR:\t\t"));
    Serial.print(radio.getSNR());   // Affichage du rapport signal sur bruit SNR
    Serial.println(F(" dB"));

    Serial.print(F("[LoRaWAN] Frequency error:\t"));
    Serial.print(radio.getFrequencyError());  // Affichage de l'erreur de fréquence
    Serial.println(F(" Hz"));

  } else if(state == RADIOLIB_ERR_RX_TIMEOUT) {
    Serial.println(F("Pas de réception!"));  // Message si aucune réception n'est reçue

  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  node.saveSession();  // Sauvegarde de la session LoRaWAN
  
  uint32_t minimumDelay = 60000;                  // Temps minimal entre l'envoi de chaque paquet (60 secondes)
  uint32_t interval = node.timeUntilUplink();     // Calcul du temps écoulé depuis la dernière transmission
  uint32_t delayMs = max(interval, minimumDelay); // Calcul du temps d'attente en fonction du délai minimal et du temps écoulé

  delay(delayMs);  // Attente avant d'envoyer un autre paquet
}

My main problem is that i use the same node to make lot of different tests and when i go back to this sketch i have to delete the node device on TTN, recreate one with the same joinEUI and devEUI then copy/paste nwkKey[] and appKey[]. It works fine like this but i would like to know if there is another method ?
Thanks
Eric

**Comment by tanguy-e** [comment]
I added these lines in the setup part just after int state = radio.begin(); to modify the default BW and SF : 
if (radio.setBandwidth(500.0) == RADIOLIB_ERR_INVALID_BANDWIDTH) {
    Serial.println(F("Selected bandwidth is invalid for this module!"));
    while (true);
  }

  if (radio.setSpreadingFactor(9) == RADIOLIB_ERR_INVALID_SPREADING_FACTOR) {
    Serial.println(F("Selected spreading factor is invalid for this module!"));
    while (true);
  }
 
But on the gateway, i have always "bandwidth": 125000 and "spreading_factor": 7,

I have someting i misunderstand?
Thanks

**Reply by HeadBoffin** [reply]
It would help to understand the context - the why you want to do this.

But mostly the join settings and subsequent interaction with the LNS will take over due to ADR.

**Comment by tanguy-e** [comment]
The context is to test. I would like to test the modification of parameters using the library and see or not the effect at the reception...

**Reply by HeadBoffin** [reply]
The LoRaWAN Network Server with ADR does all that for you. And the spec will tend to override things. You can turn off ADR but mostly you will be competing with the combined knowledge of highly experienced LNS developers.

If you want to test such things, just doing point 2 point will be far more controllable.

**Reply by StevenCellist** [reply]
ADR enabled or not, bandwidth cannot be modified within RadioLib's LoRaWAN. Only datarate, which changes SF and only in very specific cases the bandwidth.
And if you could, the gateway likely wouldn't even receive your package.

**Comment by tanguy-e** [comment]
Ok thank you very much for the clear explanation.
But i can test to increase the output power. As the default one seems to be 10 dBm for the 1262 device, i can try to increase it to 20 and see if the link values are modified ?

**Reply by HeadBoffin** [reply, marked as answer]
The LNS / ADR algorithm will adjust the power as well. It can take up to 30 uplinks for it to send a request. Depending on where you are will depend on if 20dBm is legal.

Still not sure what there is to be gained by hacking at the stack to see what happens - it's all meant to work to a spec.

**Comment by tanguy-e** [comment]
Thanks for the details
